03_Modules/EVDriver/src/module/LocalAuthListService.ts

```diff
@@ -0,0 +1,124 @@
+
+import { AttributeEnumType, SendLocalListRequest, UpdateEnumType } from "@citrineos/base";
+import { IDeviceModelRepository, ILocalAuthListRepository, VariableCharacteristics, VariableAttribute, LocalListVersion, SendLocalList, } from "@citrineos/data";
+import { LocalListAuthorization } from "@citrineos/data/src/layers/sequelize/model/Authorization/LocalListAuthorization";
+
+import { v4 as uuidv4 } from 'uuid';
+
+export class LocalAuthListService {
+  protected _localAuthListRepository: ILocalAuthListRepository;
+  protected _deviceModelRepository: IDeviceModelRepository;
+
+  constructor(localAuthListRepository: ILocalAuthListRepository,
+    deviceModelRepository: IDeviceModelRepository
+  ) {
+    this._localAuthListRepository = localAuthListRepository;
+    this._deviceModelRepository = deviceModelRepository;
+  }
+
+  /**
+   * Validates a SendLocalListRequest and persists it, then returns the correlation Id.
+   *
+   * @param {string} stationId - The ID of the station to which the SendLocalListRequest belongs.
+   * @param {string} correlationId - The correlation Id that will be used for the SendLocalListRequest.
+   * @param {SendLocalListRequest} sendLocalListRequest - The SendLocalListRequest to validate and persist.
+   * @return {SendLocalList} The persisted SendLocalList.
+   */
+  async persistSendLocalListForStationIdAndCorrelationIdAndSendLocalListRequest(stationId: string, correlationId: string, sendLocalListRequest: SendLocalListRequest): Promise<SendLocalList> {
+    const localListVersion = await this._localAuthListRepository.readOnlyOneByQuery({
+      where: {
+        stationId: stationId,
+      },
+      include: [LocalListAuthorization],
+    });
+    const sendLocalList = await this.createSendLocalListFromStationIdAndRequestAndCurrentVersion(stationId, correlationId, sendLocalListRequest, localListVersion);
+
+    const newLocalAuthListLength = await this.countUpdatedAuthListFromRequestAndCurrentVersion(sendLocalList, localListVersion);
+    // TODO If Device Model is updated to allow different variable characteristics for the same variable per station, then we need to update this
+    const maxLocalAuthListEntries = await this.getMaxLocalAuthListEntries();
+    if (!maxLocalAuthListEntries) {
+      throw new Error('Could not get max local auth list entries, required by D01.FR.12');
+    } else if (newLocalAuthListLength > maxLocalAuthListEntries) {
+      throw new Error(`Updated local auth list length (${newLocalAuthListLength}) will exceed max local auth list entries (${maxLocalAuthListEntries})`);
+    }
+
+    const itemsPerMessageSendLocalList =
+      await this.getItemsPerMessageSendLocalListByStationId(stationId)
+      || (sendLocalListRequest.localAuthorizationList ? sendLocalListRequest.localAuthorizationList?.length : 0);
+
+    if (itemsPerMessageSendLocalList && sendLocalListRequest.localAuthorizationList && itemsPerMessageSendLocalList < sendLocalListRequest.localAuthorizationList.length) {
```

Comment:
Should we make a ticket to automate this instead of throwing? Seems we could handle it in an automated fashion

Reply:
I avoided that here because of the version number. Say some upstream service pushes a SendLocalListRequest with version number 2, they may (reasonably) assume that if it's successful the version number is 2. However, if the message needed to be split up, the version number will be higher in accordance with however many messages ended up being sent.

"Hidden" functionality in the MessageAPI, like splitting up SetVariables messages in the Monitoring module, is acceptable to me when the only potential downside is partial success--here, the version number ending up out of sync could cause bigger issues and returning it in the response doesn't align with the idea that all Message API endpoints are supposed to work the same.